designs/2023-sharding/README.md

```diff
@@ -0,0 +1,140 @@
+- Repo: eslint/eslint
+- Start Date: 2023-06-01
+- RFC PR:
+- Authors: [Balavishnu V J](@balavishnuvj)
+
+# Sharding
+
+## Summary
+
+This RFC proposes adding sharding support to the ESLint package. Sharding is a technique used to divide a large workload into smaller, more manageable parts that can be processed concurrently. By introducing sharding in ESLint, we aim to improve linting performance for large codebases and reduce the overall execution time of linting tasks.
+
+## Motivation
+
+The motivation behind this feature is to address the performance limitations of ESLint when dealing with large codebases. As codebases grow in size, linting the entire codebase in a single execution can become time-consuming and impact developer productivity.
+
+By introducing sharding, ESLint can distribute the linting workload across multiple processes or threads, allowing for parallel execution and faster linting times. It can also improve integration with development tools and CI/CD pipelines, as linting tasks can be completed faster, leading to quicker feedback cycles.
+Detailed Design
+
+The implementation of sharding in the ESLint package involves the following steps:
+
+- **Configuration**: Introduce a new configuration option or flag in the ESLint configuration file to enable sharding. This option can specify the number of shards or the desired granularity of the shards.
+
+  ```
+  eslint --shard=1/3
+  ```
+
+- **Workload Division**: Analyze the input codebase and divide it into smaller units of work called shards. The division can be based on files, directories, or any other logical grouping that allows for parallel processing. Each shard should be self-contained and independent of the others. For example, larger files can be distributed across shards to balance the workload. We could also expose api to get custom division statergy.
```

Comment:
Earlier, @bradzacher [mentioned](https://github.com/eslint/rfcs/pull/112#issuecomment-1572067226):

> Sharding by rule is dangerous, in a way, because there are rules that depend on other rules to function.
>
> For example in react codebases the core no-unused-vars lint rule will not work correctly without react/jsx-uses-vars and react/jsx-uses-react from eslint-plugin-react. Those rules call context.markVariableAsUsed which in turn impacts what no-unused-vars reports!
>
> Sharding by rule would need to take these implicit dependencies into account.

To touch on this a bit more, there are various ways rules could be using shared / global state, as discussed in this [thread](https://github.com/eslint/rfcs/pull/104#discussion_r1071575733). This behavior might not be encouraged or officially supported, but just be aware that sharding even a single rule could break things if the rule gathers up state while running across files. It would be good to call this possibility out, whether rules are allowed to do this generally or only through a supported API like `markVariableAsUsed`, and how sharding could impact such rules.